Cleanup and refactoring of TrackDAO: Adding tracks [rebased and consolidated]

uklotzde opened the pull request:
TrackDAO has too many responsibilities, a multitude of hardly testable code paths, and subtle assumptions about how various member functions are supposed to interact.

As a first attempt this PR tries to straighten the code paths when adding single/multiple tracks, which is needed for the introduction of TrackCache.

uklotzde committed:
TrackDAO: Add function relocateDirectory()

Move code from a signal handler into a reusable member function.

uklotzde committed:
TrackDAO: Resolve naming conflicts with upcoming TrackCache

uklotzde committed:
Delete unused and commented code line

...the comment above explains why this does not make sense here.

uklotzde committed:
TrackDAO: Use nullptr and range-based for loops (C++11)

uklotzde committed:
TrackDAO: Rely on SoundSourceProxy for parsing track metadata

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -37,16 +36,14 @@ QMutex TrackDAO::m_sTracksMutex;
-        : m_pTrack(pTrack) {
-    DEBUG_ASSERT(m_pTrack);
+RecentTrackCacheItem::RecentTrackCacheItem(TrackPointer pTrack)
+        : m_pTrack(std::move(pTrack)) {
```
QSharedPointer has no move constructor, so std::move() is redundant. Instead we can pass the pTrack by const reference.

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -37,16 +36,14 @@ QMutex TrackDAO::m_sTracksMutex;
-    }
-    // qDebug() << "~TrackCacheItem" << m_pTrack << "ID"
+RecentTrackCacheItem::~RecentTrackCacheItem() {
+    DEBUG_ASSERT(!m_pTrack.isNull());
```
Why did you remove .._AND_HANDLE? Can't we remove the whole statement, since we check it in the constructor as well?

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -152,7 +149,7 @@ void TrackDAO::finish() {
     // directories as deleted.
     // TODO(XXX) This doesn't handle sub-directories of deleted directories.
-    foreach (QString dir, deletedHashDirs) {
+    for (const QString& dir: deletedHashDirs) {
```
Here we have decided to use "auto"

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -523,32 +520,35 @@ void TrackDAO::addTracksFinish(bool rollback) {
+            << pTrack->getLocation();
+
     int trackLocationId = -1;
     TrackId trackId;
```
I think you can reuse this invalid TrackId above and below.

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -599,14 +599,14 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
-            // weird because the track has whatever metadata the caller supplied
-            // and that metadata may differ from what is already in the
-            // database. I'm ignoring this corner case. rryan 10/2011
-            pTrack->setId(trackId);
```
Where is this call gone?

uklotzde replied:
Moved to the end of this function to avoid uncontrolled setting of the track id.

Future: Setting the track id will only be allowed from within the TrackCache 
that needs to keep its internal maps synchronized.

On 02/01/2016 12:12 AM, Daniel Schürmann wrote:
>
> In src/library/dao/trackdao.cpp 
> <https://github.com/mixxxdj/mixxx/pull/885#discussion_r51369525>:
>
> > @@ -599,14 +599,14 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
> >                  }
> >              }
> >              // Regardless of whether we unremoved this track or not -- it's
> > -            // already in the library and so we need to skip it. Set the track's
> > -            // trackId so the caller can know it. TODO(XXX) this is a little
> > -            // weird because the track has whatever metadata the caller supplied
> > -            // and that metadata may differ from what is already in the
> > -            // database. I'm ignoring this corner case. rryan 10/2011
> > -            pTrack->setId(trackId);
>
> Where is this call gone?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/885/files#r51369525>.
>

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -628,49 +628,53 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
+                << "Failed to add track to database"
+                << pTrack->getLocation();
+    }
+    DEBUG_ASSERT(pTrack->getId() == trackId);
```
this Assert here is hard to understand some comments about the possible cases would be useful.

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -682,41 +686,39 @@ QList<TrackId> TrackDAO::addTracks(const QList<QFileInfo>& fileInfoList,
                   "VALUES (:add_index, :location)");
     int index = 0;
-    foreach (const QFileInfo& rFileInfo, fileInfoList) {
+    for (const QFileInfo& fileInfo: fileInfoList) {
```
auto

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -733,25 +735,17 @@ QList<TrackId> TrackDAO::addTracks(const QList<QFileInfo>& fileInfoList,
-        QString filePath = query.value(locationColumn).toString();
-        const QFileInfo& fileInfo = fileInfoList.at(addIndex);
-        TrackPointer pTrack(new TrackInfoObject(fileInfo));
-        SoundSourceProxy(pTrack).loadTrackMetadata();
```
Where do we load now the Metadata?

daschuer replied:
Got it. I was confused because the overload of addTracksAdd()

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -1361,38 +1365,30 @@ TrackPointer TrackDAO::getTrackFromDB(TrackId trackId) const {
+    if (isTrackDirty) {
         emit(trackDirty(trackId));
+    } else {
+        emit(trackClean(trackId));
```
Why we need this signal here?

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -628,49 +628,53 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
-    addTracksPrepare();
-    addTracksAdd(pTrack, unremove);
-    addTracksFinish(false);
+    const TrackId trackId(addTracksAdd(pTrack, unremove));
```
Is this a recursive call?

daschuer replied:
No, It is an overload function. We should rename one.

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -1361,38 +1365,30 @@ TrackPointer TrackDAO::getTrackFromDB(TrackId trackId) const {
-    // If the header hasn't been parsed, parse it but only after we set the
-    // track clean and hooked it up to the track cache, because this will
-    // dirty it.
-    SoundSourceProxy(pTrack).loadTrackMetadata();
```
I do not understand why this call was here. So I am afraid I cannot tell if removing this is a good idea.
Did this always reload the track Metadata from file? Is it a regression from a previous commit? 
It should be possible to have different Metadata in Mixxx and the file. Was this broken before?

daschuer commented:
The PR now LGTM, but it is conflicting. 
The only thing I am worrying about is the bpm lock bug. Does this have dependencies to it? 
I have not yet dived into it.

uklotzde commented:
I first need to create a small follow up PR for RJ's fixes before we can 
continue to merge this PR. I also consider rebasing it on master, because I 
don't want to risk losing any of his modifications.

On 02/02/2016 09:57 PM, Daniel Schürmann wrote:
>
> The PR now LGTM, but it is conflicting.
> The only thing I am worrying about is the bpm lock bug. Does this have 
> dependencies to it?
> I have not yet dived into it.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/885#issuecomment-178815015>.
>

daschuer commented:
OK, that seams to be reasonable.

uklotzde committed:
TrackDAO: Refactor code for adding single/multiple tracks

uklotzde committed:
Use type-safe DbId for track location ids

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -2064,11 +2092,11 @@ TrackPointer TrackDAO::getOrAddTrack(const QString& trackLocation,
+    bool trackAlreadyInLibrary = trackId.isValid();
 
     // Add Track to library -- unremove if it was previously removed.
     if (!trackId.isValid()) {
```
if (!trackAlreadyInLibrary ...

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -2064,11 +2092,11 @@ TrackPointer TrackDAO::getOrAddTrack(const QString& trackLocation,
     // Add Track to library -- unremove if it was previously removed.
     if (!trackId.isValid()) {
-        trackId = addTrack(trackLocation, true);
+        trackId = addSingleTrack(trackLocation, true);
```
Here is room for optimization. We know that the track is new and we already head a TrackInfo Object inside of addSingleTrack. Can't we just add it to the cache?

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -630,49 +632,69 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
-    if (!SoundSourceProxy::isFileSupported(pTrack->getFileInfo())) {
-        // TODO(XXX) provide some kind of error code on a per-track basis.
-        return;
+    SoundSourceProxy(pTrack).loadTrackMetadata();
```
Here we load the track metadata even though we do not need it in any case. If the track is already in library it is not used. 
It is probably not a topic for this PR, but It looks like we can spit the location insert and the track insert in two functions and call the location insert before loadTrackMetadata()

daschuer commented:
It was hard to read this PR a second time ....
And it still looks good to me. I have not catch significant changes from the original version, (the first review path). Are there any? 

I have left some further improvement comments. If you are finished with this PR. I would like to merge it without these bigger changes. They can go into another PR if you or an other thinks it is worth to improve it more.   

Just give a hint when I should merge this. Thank you.

uklotzde committed:
TrackDAO: Return pointer of added track instead of only id

uklotzde committed:
TrackDAO: Don't add unremoved tracks again (fixup for edge case)

uklotzde committed:
TrackDAO: Explain why metadata needs to be loaded from file

uklotzde commented:
I may have found a minor regression that needs to be fixed: Unremoved tracks 
are marked as clean although they shouldn't. This will not increase the 
complexity substantially.

And I will also take your comments into account regarding adding a single 
track to the library: Passing the allocated TIO as the result back to the 
caller instead of throwing it away.

On 02/04/2016 06:08 PM, Daniel Schürmann wrote:
>
> It was hard to read this PR a second time ....
> And it still looks good to me. I have not catch significant changes from 
> the original version, (the first review path). Are there any?
>
> I have left some further improvement comments. If you are finished with 
> this PR. I would like to merge it without these bigger changes. They can 
> go into another PR if you or an other thinks it is worth to improve it more.
>
> Just give a hint when I should merge this. Thank you.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/885#issuecomment-179949589>.
>

uklotzde committed:
TrackDAO: Encapsulate code for caching tracks in function

uklotzde committed:
Fix regression: Notify BaseTrackCache after track has been saved

uklotzde committed:
Fix regression: Notify BaseTrackCache after track has been loaded

uklotzde committed:
Move responsibility for adding tracks to the library to TrackDAO

This fix avoids a race condition that becomes relevant when implementing
caching of tracks. The cache will not updated properly when adding track
through a bypass where the track is allocated before it is added to
the library. The cache needs to be updated atomically after a track
has been added to the library with the track id.

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -1185,6 +1221,20 @@ struct ColumnPopulator {
 
+void TrackDAO::cacheRecentTrack(
+    TrackId trackId,
+    const TrackPointer& pTrack) const {
```
Please double indent here.

daschuer commented on src/library/dao/trackdao.cpp:
```diff
@@ -630,49 +644,81 @@ bool TrackDAO::addTracksAdd(TrackInfoObject* pTrack, bool unremove) {
-        return;
+    // Initially load the metadata for the newly created track
+    // from the file.
+    SoundSourceProxy(pTrack).loadTrackMetadata();
```
If the track is already in library, this call can be skipped. 
Fixing this requires some more refactoring. So please just add a TODO

daschuer commented:
Is the track added by addSingleTrack() added to the RecentTrackCache?

uklotzde committed:
Merge branch 'master' into PrepareTrackDAOForTrackCache

uklotzde committed:
Fix indentation

uklotzde committed:
Add TODOs for future refactorings

uklotzde committed:
Add TODO for failed track addition in library scanner

uklotzde committed:
Cache single tracks after adding to the library

daschuer commented:
I have done some random tests and it still works :-) LGTM 
Thank you!
